### PR TITLE
Statement optimizations

### PR DESCRIPTION
**Pull request by tinwelint**
A couple of commits to improve some aspects of transaction handling and reading records.

### CODE REVIEW

File: community/kernel/src/main/java/org/neo4j/kernel/impl/store/CommonAbstractStore.java [hunk 1]
```diff
@@ -1089,11 +1038,36 @@ public boolean next()
 
                 try
                 {
-                    return getRecord( currentId, record, mode, pageCursor, pageIdForRecord( currentId ) ).inUse();
+                    try
+                    {
+                        record.setId( currentId );
+                        long pageId = pageIdForRecord( currentId );
+                        if ( pageId == pageCursor.getCurrentPageId() ||
+                                pageCursor.next( pageIdForRecord( currentId ) ) )
```
Comment (chrisvest) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/store/CommonAbstractStore.java]: No need to recompile pageId for the next call.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/store/CommonAbstractStore.java [hunk 2]
```diff
@@ -1089,11 +1038,36 @@ public boolean next()
...
+                        {
+                            int offset = offsetForId( currentId );
```
Comment (chrisvest) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/store/CommonAbstractStore.java]: If you place the calls to offsetForId and pageIdForRecord next to each other, and both inline, then the JIT can extract both the modulo for the offset and the division for the page id, from the same divide instruction.

File: community/kernel/src/test/java/org/neo4j/kernel/impl/api/StatementLifecycleTest.java [hunk 3]
```diff
@@ -35,14 +34,14 @@ public void shouldReleaseItselfWhenClosed() throws Exception
...
-        KernelStatement statement = new KernelStatement( transaction, null, null, null, storageStatement, new Procedures() );
+        KernelStatement statement = new KernelStatement( transaction, null, null, storageStatement, new Procedures() );
         statement.acquire();
 
         // when
         statement.close();
 
         // then
-        verify( transaction ).releaseStatement( statement );
+//        verify( transaction ).releaseStatement( statement );
```
Comment (chrisvest) [inline community/kernel/src/test/java/org/neo4j/kernel/impl/api/StatementLifecycleTest.java]: Looks like this should be fixed up

File: community/kernel/src/test/java/org/neo4j/kernel/impl/api/StatementLifecycleTest.java [hunk 4]
```diff
@@ -60,6 +59,6 @@ public void shouldReleaseWhenAllNestedStatementsClosed() throws Exception
...
```
Comment (chrisvest) [inline community/kernel/src/test/java/org/neo4j/kernel/impl/api/StatementLifecycleTest.java]: And here

**Comment by chrisvest** [comment]
Interesting stuff.

**Comment by chrisvest** [comment]
@tinwelint Really weird test failures; many about enabling upgrading.

**Comment by tinwelint** [comment]
I'll rebase once https://github.com/neo4j/neo4j/pull/6495 is in and fix the issue, because it will be much simpler then.

**Comment by systay** [comment]
@tinwelint #6495 has been merged

**Comment by tinwelint** [comment]
@systay thanks for notice

File: community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelStatement.java [hunk 5]
```diff
@@ -130,23 +151,22 @@ void assertOpen()
...
+        }
     }
 
     private boolean release()
     {
-        referenceCount--;
-
-        return (referenceCount == 0);
+        return (--referenceCount == 0);
     }
```
Comment (chrisvest) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelStatement.java]: The `acquire` and `release` methods are not symmetric with regards to their visibility, or what they do to the `storeStatement`.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactionImplementation.java [hunk 6]
```diff
@@ -169,6 +168,8 @@ public KernelTransactionImplementation( StatementOperationParts operations,
         this.pool = pool;
         this.clock = clock;
         this.tracer = tracer;
+        this.storageStatement = storeLayer.acquireStatement();
```
Comment (chrisvest) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactionImplementation.java]: I'm not seeing where this guy is closed.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/factory/GraphDatabaseFacade.java [hunk 7]
```diff
@@ -283,13 +283,16 @@ public Relationship getRelationshipById( long id )
         }
```
Comment (chrisvest) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/factory/GraphDatabaseFacade.java:283]: These ID bounds checks should be removed and left to the `StorageEngine`, so we can remove the `MAX_RELATIONSHIP_ID` and `MAX_NODE_ID` constants. The constants in `IdType` that feed our `IdGenerators` should also be removed, and we should source that data from the `RecordFormat` somehow.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/store/CommonAbstractStore.java [hunk 8]
```diff
@@ -227,7 +231,7 @@ protected void loadStorage()
         }
         catch ( IOException e )
         {
-            // TODO: Just throw IOException, add proper handling further up
+            // Just throw IOException, add proper handling further up
             throw new UnderlyingStorageException( e );
```
Comment (chrisvest) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/store/CommonAbstractStore.java]: This comment is now wrong.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/store/CommonAbstractStore.java [hunk 9]
```diff
@@ -257,34 +261,35 @@ private void extractHeaderRecord() throws IOException
...
-        byte[] data = new byte[getRecordSize()];
+        byte[] data = new byte[recordSize];
         try ( PageCursor pageCursor = storeFile.io( id / getRecordsPerPage(), PagedFile.PF_SHARED_READ_LOCK ) )
         {
             if ( pageCursor.next() )
             {
                 do
                 {
-                    pageCursor.setOffset( (int) (id % getRecordsPerPage() * getRecordSize()) );
+                    pageCursor.setOffset( (int) (id % getRecordsPerPage() * recordSize) );
```
Comment (chrisvest) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/store/CommonAbstractStore.java]: We could also add a parenthesis here.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/store/format/BaseOneByteHeaderRecordFormat.java [hunk 10]
```diff
@@ -38,72 +34,32 @@
...
-     *
-     * @param record containing data to write.
-     * @param cursor {@link PageCursor} to write the record data into.
-     * @param recordSize size of records of this format. This is passed in like this since not all formats
-     * know the record size in advance, but may be read from store header when opening the store.
-     * @throws IOException on error writing.
-     */
-    protected abstract void doWrite( RECORD record, PageCursor cursor, int recordSize, PagedFile storeFile )
-            throws IOException;
+    protected boolean isInUse( byte firstByte )
```
Comment (chrisvest) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/store/format/BaseOneByteHeaderRecordFormat.java]: I think we should have some javadocs on these methods.

File: community/kernel/src/main/java/org/neo4j/storageengine/api/StorageStatement.java [hunk 11]
```diff
@@ -95,6 +95,8 @@
      */
     Cursor<RelationshipItem> relationshipsGetAllCursor();
 
+    void acquire();
```
Comment (chrisvest) [inline community/kernel/src/main/java/org/neo4j/storageengine/api/StorageStatement.java]: Please add javadocs that describe what this is supposed to mean.

**Comment by chrisvest** [comment]
@tinwelint I've done my first review.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactionImplementation.java (hunk 6 above)
Comment (tinwelint) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactionImplementation.java]: Since that method no longer acquires a statement, but merely creates one it should be renamed to `newStatement()`. Calling `acquire()` acquires it. It's quite internal and KTI is basically the only user so that slight inconvenience of this change is minimal btw. I'll do this rename.

... and so that guy shouldn't be closed by KTI, but will actually will be by KernelStatement.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/factory/GraphDatabaseFacade.java (hunk 7 above)
Comment (tinwelint) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/factory/GraphDatabaseFacade.java:283]: Right, that's true. Shall we leave that to another PR perhaps?

File: community/kernel/src/main/java/org/neo4j/kernel/impl/store/CommonAbstractStore.java (hunk 9 above)
Comment (tinwelint) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/store/CommonAbstractStore.java]: We could use the `RecordPageLocationCalculator` here as well.

**Comment by tinwelint** [comment]
@chrisvest comments addressed.

**Comment by chrisvest** [comment]
@tinwelint Nice, but now it needs a rebase.
